src/QsCompiler/DataStructures/SymbolInformation.fs

```diff
@@ -0,0 +1,78 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.Quantum.QsCompiler.SyntaxTree
+
+open Microsoft.Quantum.QsCompiler.DataTypes
+
+/// used to represent the names of declared type parameters or the name of the declared argument items of a callable
+type QsLocalSymbol =
+    | ValidName of string
+    | InvalidName
+
+/// used to represent information on typed expressions generated and/or tracked during compilation
+type InferredExpressionInformation =
+    {
+        /// whether or not the value of this expression can be modified (true if it can)
+        IsMutable: bool
+        /// indicates whether the annotated expression directly or indirectly depends on an operation call within the surrounding implementation block
+        /// -> it will be set to false for variables declared within the argument tuple
+        /// -> using and borrowing are *not* considered to implicitly invoke a call to an operation, and are thus *not* considered to have a quantum dependency.
+        HasLocalQuantumDependency: bool
+    }
+
+    /// Returns the inferred expression information used for declared parameters to callables.
+    /// Note that this information is used within the declaration but not for the argument(s) in any given call-like expressions.
+    static member ParameterDeclaration = { IsMutable = false; HasLocalQuantumDependency = false }
+
+type LocalVariableDeclaration<'Name, 'Type> =
+    {
+        /// the name of the declared variable
+        VariableName: 'Name
+        /// the fully resolved type of the declared variable
+        Type: 'Type
```

Comment:
It is not clear to me yet why the type of the `Type` member has to be a type-parameter. I see that this type-parameter propagates from the `Lambda` type so maybe the question is, why `Lambda` needs a type-parameter? Why isn't `ResolvedType` enough?

Reply:
We make a type distinction between the type checked data structures and the unchecked data structures. I think this distinction between validated AST nodes vs unvalidated one is useful. At the same time, rather than a full duplication of all data structures, some structures are reused, and in particular the major DUs like ExpressionKind and TypeKind. These DUs are hence type parameterized to reflect when they are type checked and when they are not. The choice to reuse the major DUs is debatable, but there are benefits and downsides and it is the current setup. With the lambda data structure being part of such a dual-purpose DU, the same is done here for the lambda; the type here reflects and is different depending on whether the lambda has been validated or not.

Reply:
Thank you for the explanation! It makes sense now.